hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+import com.hedera.services.state.merkle.MerkleAccount;
+import com.hedera.services.state.merkle.MerkleNetworkContext;
+import com.hedera.services.state.merkle.MerkleStakingInfo;
+import com.hedera.services.utils.EntityNum;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.swirlds.merkle.map.MerkleMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.time.LocalDate;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.hedera.services.context.properties.StaticPropertiesHolder.STATIC_PROPERTIES;
+import static com.hedera.services.ledger.accounts.staking.RewardCalculator.zoneUTC;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.finalBalanceGiven;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.hasStakeFieldChanges;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.isWithinRange;
+import static com.hedera.services.ledger.properties.AccountProperty.DECLINE_REWARD;
+
+public class StakeAwareAccountsCommitsInterceptor extends AccountsCommitInterceptor {
+	private final StakeChangeManager manager;
+	private final Supplier<MerkleNetworkContext> networkCtx;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo;
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final RewardCalculator rewardCalculator;
+	private final SideEffectsTracker sideEffectsTracker;
+	private final GlobalDynamicProperties dynamicProperties;
+
+	private boolean rewardsActivated;
+	private boolean rewardBalanceChanged;
+	private long newRewardBalance;
+
+	private static final Logger log = LogManager.getLogger(StakeAwareAccountsCommitsInterceptor.class);
+	private static final long STAKING_FUNDING_ACCOUNT_NUMBER = 800L;
+
+	public StakeAwareAccountsCommitsInterceptor(
+			final SideEffectsTracker sideEffectsTracker,
+			final Supplier<MerkleNetworkContext> networkCtx,
+			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo,
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final RewardCalculator rewardCalculator,
+			final StakeChangeManager manager) {
+		super(sideEffectsTracker);
+		this.manager = manager;
+		this.networkCtx = networkCtx;
+		this.stakingInfo = stakingInfo;
+		this.accounts = accounts;
+		this.rewardCalculator = rewardCalculator;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.dynamicProperties = dynamicProperties;
+	}
+
+	@Override
+	public void preview(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		// if the rewards are activated previously they will not be activated again
+		rewardsActivated = rewardsActivated || networkCtx.get().areRewardsActivated();
+		rewardBalanceChanged = false;
```

Comment:
We can just set `newStakingRewardBalance = -1` and then update it, if appropriate, at the end of `preview()`.